include/zephyr/bluetooth/classic/a2dp_codec_sbc.h

```diff
@@ -0,0 +1,124 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef ZEPHYR_INCLUDE_BLUETOOTH_A2DP_CODEC_H_
+#define ZEPHYR_INCLUDE_BLUETOOTH_A2DP_CODEC_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* Sampling Frequency */
+#define A2DP_SBC_SAMP_FREQ_16000 BIT(7)
+#define A2DP_SBC_SAMP_FREQ_32000 BIT(6)
+#define A2DP_SBC_SAMP_FREQ_44100 BIT(5)
+#define A2DP_SBC_SAMP_FREQ_48000 BIT(4)
+
+/* Channel Mode */
+#define A2DP_SBC_CH_MODE_MONO  BIT(3)
+#define A2DP_SBC_CH_MODE_DUAL  BIT(2)
+#define A2DP_SBC_CH_MODE_STREO BIT(1)
+#define A2DP_SBC_CH_MODE_JOINT BIT(0)
+
+/* Block Length */
+#define A2DP_SBC_BLK_LEN_4  BIT(7)
+#define A2DP_SBC_BLK_LEN_8  BIT(6)
+#define A2DP_SBC_BLK_LEN_12 BIT(5)
+#define A2DP_SBC_BLK_LEN_16 BIT(4)
+
+/* Subbands */
+#define A2DP_SBC_SUBBAND_4 BIT(3)
+#define A2DP_SBC_SUBBAND_8 BIT(2)
+
+/* Allocation Method */
+#define A2DP_SBC_ALLOC_MTHD_SNR      BIT(1)
+#define A2DP_SBC_ALLOC_MTHD_LOUDNESS BIT(0)
+
+#define BT_A2DP_SBC_SAMP_FREQ(cap)    ((cap->config[0] >> 4) & 0x0f)
+#define BT_A2DP_SBC_CHAN_MODE(cap)    ((cap->config[0]) & 0x0f)
+#define BT_A2DP_SBC_BLK_LEN(cap)      ((cap->config[1] >> 4) & 0x0f)
+#define BT_A2DP_SBC_SUB_BAND(cap)     ((cap->config[1] >> 2) & 0x03)
+#define BT_A2DP_SBC_ALLOC_MTHD(cap)   ((cap->config[1]) & 0x03)
+
+/** @brief SBC Codec */
+struct bt_a2dp_codec_sbc_params {
+	/** First two octets of configuration */
+	uint8_t config[2];
+	/** Minimum Bitpool Value */
+	uint8_t min_bitpool;
+	/** Maximum Bitpool Value */
+	uint8_t max_bitpool;
+} __packed;
+
+struct bt_a2dp_codec_sbc_media_packet_hdr {
+#ifdef CONFIG_LITTLE_ENDIAN
+	/** If the F bit is set to 0, this field indicates the number of frames contained
+	 *  in this packet. If the F bit is set to 1, this field indicates the number
+	 *  of remaining fragments, including the current fragment.
+	 *  Therefore, the last counter value shall be one.
+	 */
+	uint8_t number_of_sbc_frames:4;
+	/** Reserved */
+	uint8_t rfa:1;
+	/** Set to 1 for the last packet of a fragmented SBC frame, otherwise set to 0. */
+	uint8_t L:1;
+	/** Set to 1 for the starting packet of a fragmented SBC frame, otherwise set to 0. */
+	uint8_t S:1;
+	/** Set to 1 if the SBC frame is fragmented, otherwise set to 0. */
+	uint8_t F:1;
+#else
+	/** Set to 1 if the SBC frame is fragmented, otherwise set to 0. */
+	uint8_t F:1;
+	/** Set to 1 for the starting packet of a fragmented SBC frame, otherwise set to 0. */
+	uint8_t S:1;
+	/** Set to 1 for the last packet of a fragmented SBC frame, otherwise set to 0. */
+	uint8_t L:1;
+	/** Reserved */
+	uint8_t rfa:1;
+	/** If the F bit is set to 0, this field indicates the number of frames contained
+	 *  in this packet. If the F bit is set to 1, this field indicates the number
+	 *  of remaining fragments, including the current fragment.
+	 *  Therefore, the last counter value shall be one.
+	 */
+	uint8_t number_of_sbc_frames:4;
```

Comment:
It really isn't ideal that you're ending up duplicating all of the documentation too. Every time you need to fix or update something you will then need to do it in two places. I also don't think this is particularly intuitive for someone who primarily reads documentation from the code rather than something generated by oxygen. I'd consider refactoring out the documentation to a single block (e.g. everything before the struct), but really I'm starting to regret giving you flexibility to choose this path rather than using helper macros :) Would it really be that bad to have something like this:
```c
#define BT_A2DP_SBC_MEDIA_HDR_NUM_FRAMES(hdr) FIELD_GET(GENMASK(3, 0), (hdr))
#define BT_A2DP_SBC_MEDIA_HDR_L(hdr)          FIELD_GET(BIT(5), (hdr))
#define BT_A2DP_SBC_MEDIA_HDR_S(hdr)          FIELD_GET(BIT(6), (hdr))
#define BT_A2DP_SBC_MEDIA_HDR_F(hdr)          FIELD_GET(BIT(7), (hdr))
```

Reply:
Appreciate your comments.
My main concern is that the application will use the struct fields too. like the follow codes in shell\a2dp.c.
```
	sbc_hdr = net_buf_add(buf, sizeof(struct bt_a2dp_codec_sbc_media_packet_hdr));
	memset(sbc_hdr, 0, sizeof(struct bt_a2dp_codec_sbc_media_packet_hdr));
	sbc_hdr->number_of_sbc_frames = 1;
```
If we use the MACRO, then I think we need to add GET/SET macros as follow, and application need to use the MACRO  (from my viewpoint, it is not convenient).
```
#define BT_A2DP_SBC_MEDIA_HDR_NUM_FRAMES_GET(hdr) FIELD_GET(GENMASK(3, 0), (hdr)) //hdr is value.
#define BT_A2DP_SBC_MEDIA_HDR_NUM_FRAMES_SET(hdr, val) hdr = (hdr & ~GENMASK(3, 0)) | (GENMASK(3, 0) & val)
```
What do you think?

Reply:
Looks alright, but I think it'd make sense to use `FIELD_PREP()` for the setting of the value. It might also make sense to encode the entire header in one go, i.e. have the setter macro take all fields as separate parameters and then construct the unified uint8_t out of them, something like:

`*hdr = BT_A2DP_SBC_MEDIA_HDR_ENCODE(num_frames, l, s, f);`

Reply:
Codes are updated. I add xx_GET, xx_SET and xx_ENCODE.